File: Models/Template.cs
namespace ranking_app.Models;

public class TemplateModel
{
    public int Id { get; set; }
    public required string Name { get; set; }
    public required string UserId { get; set; }
    public ICollection<Tier> Tiers { get; set; }
    public ICollection<Element> Elements { get; set; }
    public DateTime CreatedAt { get; set; }
}

public class TemplatePutPayload : TemplateModel
{
    public ICollection<TierPayload> Tiers { get; set; }
    public required ICollection<ElementPayload> Elements { get; set; }
}

File: ClientApp/src/models/Template.ts
import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
